x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java

```diff
@@ -0,0 +1,155 @@
+
+        final EcsLayout ecsLayout = ECSJsonLayout.newBuilder().setType("deprecation").setConfiguration(configuration).build();
+
+        this.filter = new RateLimitingFilter();
+        this.appender = new DeprecationIndexingAppender("deprecation_indexing_appender", filter, ecsLayout, consumer);
+    }
+
+    @Override
+    protected void doStart() {
+        this.appender.start();
+        Loggers.addAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+    }
+
+    @Override
+    protected void doStop() {
+        Loggers.removeAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+        this.appender.stop();
+    }
+
+    @Override
+    protected void doClose() {
+        this.processor.close();
+    }
+
+    /**
+     * Listens for changes to the cluster state, in order to know whether to toggle indexing
+     * and to set the cluster UUID and node ID. These can't be set in the constructor because
+     * the initial cluster state won't be set yet.
+     *
+     * @param event the cluster state event to process
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        final boolean newEnabled = WRITE_DEPRECATION_LOGS_TO_INDEX.get(state.getMetadata().settings());
+        if (appender.isEnabled() != newEnabled) {
+            // We've flipped from disabled to enabled. Make sure we start with a clean cache of
+            // previously-seen keys, otherwise we won't index anything.
+            if (newEnabled) {
+                this.filter.reset();
+            }
+            appender.setEnabled(newEnabled);
+        }
+    }
+
+    /**
+     * Constructs a {@link Consumer} that knows what to do with the {@link IndexRequest} instances that the
+     * {@link DeprecationIndexingAppender} creates. This logic is separated from the service in order to make
+     * testing significantly easier, and to separate concerns.
+     * <p>
+     * Writes are done via {@link BulkProcessor}, which handles batching up writes and retries.
+     *
+     * @param threadPool due to <a href="https://github.com/elastic/elasticsearch/issues/50440">#50440</a>,
+     *                   extra care must be taken to avoid blocking the thread that writes a deprecation message.
+     * @return           a consumer that accepts an index request and handles all the details of writing it
+     *                   into the cluster
+     */
+    private Consumer<IndexRequest> buildIndexRequestConsumer(ThreadPool threadPool) {
+        return indexRequest -> {
+            try {
+                // TODO: remove the threadpool wrapping when the .add call is non-blocking
+                // (it can currently execute the bulk request occasionally)
+                // see: https://github.com/elastic/elasticsearch/issues/50440
+                threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> this.processor.add(indexRequest));
```

Comment:
`add` can indeed block, but only when the max concurrent requests are met and when the number of documents or size of docs are are exceeded (as opposed to the timed flush). By default the number of concurrent requests is 1, document count is 1000 and size is 5MB... so if any of these conditions are met by default within the default flush interval of 10s it will block on add. 

Rather forking here I would suggest to tweak the bulk processor to avoid blocking.

The rationale here is that if the for some reason the index is not writable for a very long period of time, you can end up with a large number of threads from the GENERIC pool. This is further compounded by the fact that some (409) write exceptions are retried by default, leaving the thread alive 5s+ by default.  I understand this is protected by the deprecation throttling..so maybe it is moot point once things are warmed up. 

I 100% agree with the motivation here to ensure that we avoid blocking here, however I think we can better address this by 1) increasing the concurrent requests for the bulk processor (maybe = core count with min 2 ?) 2) disable the bulk processor retry, or configure the retry to be less forgiving (as to consume a concurrent request slot for less time) 3) bump the document count and size default up high so that the flush scheduler will be much more likely to be thing blocks if anything blocks. 4) (optionally) add a new configuration that will not try to execute a request if the max concurrent requests are met. eliminating the potential to even block the flush thread.

Reply:
I've pushed some changes in e8429ab2000, please let me know if this is what you were intending. Since `BulkProcessor` apparently already supports disabling the size and doc count checks, I went ahead and did that, and tweaked the backoff times along the same lines as an ILM backoff I found, the total retry time should be less I think?

Reply:
yes, this is what I was intending... an additional commentent inline: https://github.com/elastic/elasticsearch/pull/61484#discussion_r481431419